### PR TITLE
hwdef: Call `str.startswith()` once with a tuple

### PR DESCRIPTION
**Pull request by cclauss**
% `ruff rule PIE810`
> The `startswith` and `endswith` methods accept tuples of prefixes or suffixes respectively.
Passing a tuple of prefixes or suffixes is more efficient and readable than calling the method multiple times.

> References
- [Python documentation: `str.startswith`](https://docs.python.org/3/library/stdtypes.html#str.startswith)
- [Python documentation: `str.endswith`](https://docs.python.org/3/library/stdtypes.html#str.endswith)
---
% `ruff check --select=PIE --ignore=PIE790 --output-format=concise libraries/AP_HAL_ChibiOS`
```
libraries/AP_HAL_ChibiOS/hwdef/scripts/chibios_hwdef.py:154:12: PIE810 Call `endswith` once with a `tuple`
libraries/AP_HAL_ChibiOS/hwdef/scripts/chibios_hwdef.py:167:26: PIE810 Call `endswith` once with a `tuple`
libraries/AP_HAL_ChibiOS/hwdef/scripts/chibios_hwdef.py:168:17: PIE810 Call `startswith` once with a `tuple`
libraries/AP_HAL_ChibiOS/hwdef/scripts/chibios_hwdef.py:771:12: PIE810 Call `startswith` once with a `tuple`
libraries/AP_HAL_ChibiOS/hwdef/scripts/chibios_hwdef.py:801:14: PIE810 Call `startswith` once with a `tuple`
libraries/AP_HAL_ChibiOS/hwdef/scripts/chibios_hwdef.py:806:15: PIE810 Call `startswith` once with a `tuple`
libraries/AP_HAL_ChibiOS/hwdef/scripts/chibios_hwdef.py:1553:17: PIE810 Call `startswith` once with a `tuple`
libraries/AP_HAL_ChibiOS/hwdef/scripts/chibios_hwdef.py:1562:16: PIE810 Call `startswith` once with a `tuple`
libraries/AP_HAL_ChibiOS/hwdef/scripts/chibios_hwdef.py:2398:16: PIE810 Call `startswith` once with a `tuple`
libraries/AP_HAL_ChibiOS/hwdef/scripts/chibios_hwdef.py:2674:16: PIE810 Call `startswith` once with a `tuple`
libraries/AP_HAL_ChibiOS/hwdef/scripts/dma_resolver.py:104:10: PIE810 Call `startswith` once with a `tuple`
libraries/AP_HAL_ChibiOS/hwdef/scripts/dma_resolver.py:270:13: PIE810 Call `startswith` once with a `tuple`
Found 12 errors.
```

### CODE REVIEW

File: libraries/AP_HAL_ChibiOS/hwdef/scripts/dma_resolver.py [hunk 1]
```diff
@@ -169,7 +169,7 @@ def generate_DMAMUX_map_mask(peripheral_list, channel_mask, noshare_list, dma_ex
         base = idx % 16
         for i in range(16):
             found = None
-            for ii in list(range(base,16)) + list(range(0,base)):
```
Comment (amilcarlucas) [inline libraries/AP_HAL_ChibiOS/hwdef/scripts/dma_resolver.py]: 0,base is a drive-by?

Comment (cclauss) [reply libraries/AP_HAL_ChibiOS/hwdef/scripts/dma_resolver.py]: Please read the commit message when reviewing pull requests.
> libraries/AP_HAL_ChibiOS/hwdef/scripts/dma_resolver.py:172:57: PIE808 [*] Unnecessary `start` argument in `range`

https://docs.astral.sh/ruff/rules/unnecessary-range-start

Comment (tpwrules) [reply libraries/AP_HAL_ChibiOS/hwdef/scripts/dma_resolver.py]: That information is not in the commit message? The commit message doesn’t have any of the PR text contents either. Please split that change into a separate commit or PR.

I also like the symmetry here and wouldn’t propose changing it.

Comment (cclauss) [reply libraries/AP_HAL_ChibiOS/hwdef/scripts/dma_resolver.py]: Reverted and commit message edited.

**Review by peterbarker — COMMENTED** [review]
This looks good.

How was it tested?

**Comment by cclauss** [comment]
Step one is to read the two links to the Python documentation provided in the commit message.  They explain that these transformations to reduce unnecessary repetitive function calls have been recommended since 2006 when [they were introduced](https://docs.python.org/2/whatsnew/2.5.html#other-language-changes) in Python 2.5

The original Python implementation and tests can be found in:
* https://github.com/sbdchd/flake8-pie/pull/121/files

In the Ruff rewrite, astral-sh/ruff#2616 was implemented in Rust at:
https://github.com/astral-sh/ruff/blob/5e02d839d5778a0faae468b196d7b320efdbddbe/crates/ruff_linter/src/rules/flake8_pie/rules/multiple_starts_ends_with.rs#L70

The Python test fixture is at:
https://github.com/astral-sh/ruff/blob/main/crates/ruff_linter/resources/test/fixtures/flake8_pie/PIE810.py

These tests are executed at:
https://github.com/astral-sh/ruff/blob/5e02d839d5778a0faae468b196d7b320efdbddbe/crates/ruff_linter/src/rules/flake8_pie/mod.rs#L17

Issues that have been used to handle corner cases and add more tests include:
* astral-sh/ruff#3360
* astral-sh/ruff#3515
* astral-sh/ruff#4323
* astral-sh/ruff#8906
* astral-sh/ruff#9663

PIE810 has been tested by being linted for in:
NumPy: https://github.com/numpy/numpy/blob/70851b385201e12778413df9c5e0b1df31268c35/ruff.toml#L30
pandas: https://github.com/pandas-dev/pandas/blob/c067bcd701e6cb4125e869a2802ef867d8395800/pyproject.toml#L228
Poetry: https://github.com/python-poetry/poetry/blob/d8bc3610807aa15d4dcd9000a2df95b9c0af3908/pyproject.toml#L124
Pydantic: https://github.com/pydantic/pydantic/blob/0c4a22b64b23dfad27387750cf07487efc45eb05/pyproject.toml#L210
emscripten: https://github.com/emscripten-core/emscripten/blob/6db95b6147b33b545c1738c131face0fb9e5f2cc/pyproject.toml#L28
pyscript: https://github.com/pyscript/pyscript/blob/14cc05fb8072693411fd373539a4095fffdbde0f/pyproject.toml#L7
pytest: https://github.com/pytest-dev/pytest/blob/336cb917f7175ccc5a699f81b58f3587c5a7dcca/pyproject.toml#L106
openlibrary: https://github.com/internetarchive/openlibrary/blob/8ee01197ccb6c28c6f466e267a5cdf4269eaeff1/pyproject.toml#L77
and many other Python codebases: https://github.com/astral-sh/ruff#whos-using-ruff

**Comment by peterbarker** [comment]
> Step one is to read the two links to the Python documentation provided in the commit message. They explain that these transformations to reduce unnecessary repetitive function calls have been recommended since 2006 when [they were introduced](https://docs.python.org/2/whatsnew/2.5.html#other-language-changes) in Python 2.5
> 
> The original Python implementation and tests can be found in:

I was not asking how the tools used to make the transformations were tested.

I was asking how the scripts modified in this PR were tested after being modified.

**Comment by cclauss** [comment]
Is there no test coverage for this code?  Given the code complexity, tests should exist.
% `ruff check --select=C9,PLR09 --output-format=concise libraries/AP_HAL_ChibiOS/hwdef/scripts/chibios_hwdef.py`

**Comment by Ryanf55** [comment]
> Is there no test coverage for this code? Given the code complexity, tests should exist. % `ruff check --select=C9,PLR09 --output-format=concise libraries/AP_HAL_ChibiOS/hwdef/scripts/chibios_hwdef.py`

Unfortunately, very little automated testing exists for ChibiOS specific components. A dev team member can recommend a test plan, and I can help run it.
